data/tokenizer.py

```diff
@@ -0,0 +1,238 @@
+            if token == "<pad>":
+                break
+            if idx % 3 == 0:
+                # pitch token
+                buff.clear()
+                # [:-1] to clear p letter from a token
+                buff.append(eval(token[:-1]))
+            elif idx % 3 == 1:
+                # time token
+                buff.append(eval(txt) for txt in token.split("-"))
+            else:
+                # velocity token
+                # [:-1] to clear v letter from the token
+                buff.append(eval(token[:-1]))
+                samples.append(buff)
+
+        df = pd.DataFrame(samples, columns=self.keys)
+
+        return df
+
+    def untokenize_tgt(self, tokens: list[str]) -> list[int]:
+        # [:-1] to remove v letter from the end of a token
+        velocities = [int(token[:-1]) for token in tokens if token not in self.specials]
+
+        return velocities
+
+
+class MultiStartEncoder(MultiTokEncoder):
+    def __init__(self, quantization_cfg: DictConfig, keys: list[str] = None, tgt_bins: int = 650):
+        super().__init__()
+        self.quantization_cfg = quantization_cfg
+        self.tgt_bins = tgt_bins
+        if keys is None:
+            self.keys = ["pitch", "dstart_bin", "duration_bin", "velocity_bin"]
+        else:
+            self.keys = keys
+
+        self.specials = ["<CLS>", "<pad>"]
+
+        self.vocab = list(self.specials)
+
+        # add midi tokens to vocab
+        self._build_vocab()
+        self.token_to_id = {token: it for it, token in enumerate(self.vocab)}
+
+    def __rich_repr__(self):
+        yield "MultiStartEncoder"
+        yield "vocab_size", self.vocab_size
+
+    @property
+    def vocab_size(self) -> int:
+        return len(self.vocab)
+
+    def _build_vocab(self):
+        time_tokens_product = itertools.product(
+            # weird because we want to use dstart if there is a dstart_bin key, and start if there is a start_bin key
+            range(self.quantization_cfg[self.keys[1][:-4]]),
+            range(self.quantization_cfg.duration),
+        )
+
+        for pitch in range(21, 109):
+            self.vocab.append(f"{pitch}p")
+        for start, duration in time_tokens_product:
+            self.vocab.append(f"{start:0.0f}-{duration:0.0f}")
+        for velocity in range(self.quantization_cfg.velocity):
+            self.vocab.append(f"{velocity:0.0f}v")
+        for start in range(self.tgt_bins):
+            self.vocab.append(f"{start}s")
+
+    def tokenize_src(self, record: dict) -> list[str]:
+        tokens = []
+        n_samples = len(record[self.keys[0]])
+        for idx in range(n_samples):
+            # append p and v so that there are no duplicate tokens
+            pitch_token = f"{record[self.keys[0]][idx]:0.0f}p"
+            time_token = f"{record[self.keys[1]][idx]}-{record[self.keys[2]][idx]:0.0f}"
+            velocity_token = f"{record[self.keys[3]][idx]:0.0f}v"
+            tokens += [pitch_token, time_token, velocity_token]
```

Comment:
```suggestion
            pitch_token = f"PITCH-{record[self.keys[0]][idx]:0.0f}"
            time_token = f"TIME-{record[self.keys[1]][idx]}-{record[self.keys[2]][idx]:0.0f}"
            velocity_token = f"VEL-{record[self.keys[3]][idx]:0.0f}"
            tokens += [pitch_token, time_token, velocity_token]
```

I think there's no technical reason to keep those short, and I suspect we'll have to look at tokens occasionally, so it's better to make them more explicit.